source/client-side-operations-timeout/client-side-operations-timeout.rst

```diff
@@ -0,0 +1,875 @@
+undesirable behavior because it does not match the guarantees made by
+``timeoutMode=ITERATION``. To work around this, drivers honor ``timeoutMS``
+for both the original operation and all ``getMore``’s but only use it to
+derive client-side timeouts and do not append a ``maxTimeMS`` field to any
+commands. The server-side execution time is enforced via socket timeouts.
+
+The second type is tailable awaitData cursors. The server supports the
+``maxTimeMS`` option for the original command. For ``getMore``’s, the option
+is supported, but instead of limiting the server-side execution time, it
+specifies how long the server should wait for new data to arrive if it
+reaches the end of the capped collection and the batch is still empty. If no
+new data arrives within that time limit, the server will respond with an
+empty batch. For these cursors, drivers support both the ``timeoutMS`` and
+``maxAwaitTimeMS`` options. The ``timeoutMS`` option is used to derive
+client-side timeouts, while the ``maxAwaitTimeMS`` option is used as the
+``maxTimeMS`` field for ``getMore`` commands. These values have distinct
+meanings, so supporting both yields a more robust, albeit verbose, API.
+Drivers error if ``maxAwaitTimeMS`` is greater than or equal to ``timeoutMS``
+because in that case, ``getMore`` requests would not succeed if the batch was
+empty: the server would wait for ``maxAwaitTimeMS``, but the driver would
+close the socket after ``timeoutMS``.
+
+Change stream behavior
+----------------------
+
+Change streams internally behave as tailable awaitData cursors, so the
+behavior of the ``timeoutMS`` option is the same for both. The main
+difference is that change streams are resumable and drivers automatically
+perform resume attempts when they encounter transient errors. This allows
+change streams to be resilient to timeouts. If ``timeoutMS`` expires during a
+next call, drivers can’t auto-resume, but they can make sure the change
+stream is not invalidated so the user can call next again. In this case, the
+subsequent call would perform the resume without doing a ``getMore`` first.
+
+withTransaction communicates ``timeoutMS`` via ClientSession
+--------------------------------------------------------
+
+Because the ``withTransaction`` API doesn’t allow drivers to plumb down the
+remaining timeout into the user-provided callback, this spec requires the
+remaining timeout to be stored on the ClientSession. Operations in the
+callback that run under that ClientSession can then extract the timeout from
+the session and apply it. To avoid confusing validation semantics, operations
+error if there is a timeout on the session but also an overridden timeout for
+the operation. It’s possible that the ability to communicate timeouts for a
+block of operations via a ClientSession is useful as a general purpose API,
+but we’ve decided to make it private until there are other known use cases.
```

Comment:
In an earlier comment I talk about potentially using a ClientSession to share the calculated timeout between libmongoc and PHPLIB (userland), which would no doubt require PHPC (C extension) to provide a public API. Does this count as a "known use case" that would warrant standardizing such an API or should be just create that independently for our own needs?

Reply:
I don't think of this as a known use case. In my opinion, a known use case would be something like "I want to execute a block of operations under the same explicit session and have them all execute within X milliseconds, but don't want to use a transaction". This sort of functionality would require a spec change to provide a cross-drivers API.

I'm not familiar with PHPC extensions and the communication patterns between libmongoc and PHPLIB, though. Are you proposing that libmongoc expose an API to communicate timeouts via sessions that PHP would then use to implement to implement this project? If so, would this API be available to all users?

Reply:
> Are you proposing that libmongoc expose an API to communicate timeouts via sessions that PHP would then use to implement to implement this project? If so, would this API be available to all users?

Yes. A client session is presently the only object that's accessible throughout the entire stack from libmongoc through PHPC to PHPLIB (and vice versa).

This doesn't seem like a jump given that "Convenient Transactions API" already talks about tracking the timeout on a session:

> To propagate the timeout to the user-supplied callback, drivers MUST store the timeout as a field on the ClientSession object. This field SHOULD be private to ensure that a user can not modify it while a withTransaction call is in progress. Drivers that cannot make this field private MUST signal that the field should not be accessed or modified by users if there is an idiomatic way to do so in the language (e.g. underscore-prefixed variable names in Python) and MUST document that modification of the field can cause unintended correctness issues for applications.

In our case, PHPLIB can only utilize the public API of PHPC. In turn, PHPC can only utilize the public API of libmongoc. There's be no way to restrict users from accessing this.

Your explanation that this does not constitute of "known use case" (as far as this spec is concerned) makes sense, though.

Reply:
This seems like a very C/PHP-specific implementation detail, so I don't feel it's worth it to add general-purpose guidance to the spec to address it. I'll resolve this conversation and authors can use their discretion to determine a good design for this feature. Happy to weigh in during those implementations offline if additional guidance is needed, though.